checks/raw/sast_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package raw
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/ossf/scorecard/v4/checker"
+	"github.com/ossf/scorecard/v4/clients"
+	mockrepo "github.com/ossf/scorecard/v4/clients/mockclients"
+	"github.com/ossf/scorecard/v4/finding"
+)
+
+func TestSAST(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name     string
+		files    []string
+		commits  []clients.Commit
+		expected checker.SASTData
+	}{
+		{
+			name: "has codeql 1",
+			files: []string{
+				".github/workflows/workflow-not-pinned.yaml",
+				".github/workflows/pom.xml",
+			},
+			commits: []clients.Commit{
+				{
+					AssociatedMergeRequest: clients.PullRequest{
+						Number: 1,
+					},
+				},
+			},
+			expected: checker.SASTData{
+				Workflows: []checker.SASTWorkflow{
+					{
+						Type: checker.CodeQLWorkflow,
+						File: checker.File{
+							Path:   ".github/workflows/workflow-not-pinned.yaml",
+							Offset: checker.OffsetDefault,
+							Type:   finding.FileTypeSource,
+						},
+					},
+					{
+						Type: checker.SonarWorkflow,
+						File: checker.File{
+							Path:      ".github/workflows/pom.xml",
+							Type:      finding.FileTypeSource,
+							Snippet:   "https://sonarqube.private.domain",
+							Offset:    2,
+							EndOffset: 2,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:  "has codeql 2",
+			files: []string{".github/workflows/github-workflow-multiple-unpinned-uses.yaml"},
```

Comment:
What's the difference between the test workflows in `./testdata` and the test workflows in `.github` at repo root?

Reply:
They are identical. I copy pasted to make the file pathing easy for multiple test files in the same test.

Reply:
Ah, is the top-level `.github` for test workflows that can be used across multiple packages, while the local `testdata` is for test workflows solely for a particular package's tests?

Reply:
> Ah, is the top-level `.github` for test workflows that can be used across multiple packages, while the local `testdata` is for test workflows solely for a particular package's tests?

The reason was so I could look up files like this: 
https://github.com/ossf/scorecard/pull/3571/files/8a05a16cd9ef8bd05526acebfd52dcdde864ee1e#diff-ee1cd92093aa2169a9a3b1528d3e99b83eca55041cdeeaa9eba8a2a9d75a5bb2R141-R148

The reason I need that is that the filename should begin with `.github/workflow` for this part:
https://github.com/ossf/scorecard/pull/3571/files/8a05a16cd9ef8bd05526acebfd52dcdde864ee1e#diff-ac5922dc86cd2646795a8479719a157b0921e3a80a15eb6ccf34d60602dd098bR125-R127

... which is taken as-is from the previous check.

Should we avoid having duplication in test files?

Reply:
Closing since there doesn't seem to be an issue. Feel free to re-open if that is not the case.